src/pepr/operator/controllers/network/generators/kubeNodes.ts

```diff
@@ -0,0 +1,174 @@
+  }
+}
+
+/**
+ * Returns the egress CIDRs of all known nodes as network policy peers.
+ * If none are known, defaults to 0.0.0.0/0 and logs a warning.
+ */
+export function kubeNodes(): V1NetworkPolicyPeer[] {
+  const policies = buildNodePolicies([...nodeSet]);
+  if (policies.length > 0) return policies;
+
+  log.warn("Unable to get Node CIDRs, defaulting to 0.0.0.0/0");
+  return [anywhere];
+}
+
+/**
+ * When a node is created or updated, if it's Ready, add its IP to the set,
+ * rebuild the policies, and update the NetworkPolicies.
+ */
+export async function updateKubeNodesFromCreateUpdate(node: kind.Node) {
+  const isReady = node.status?.conditions?.some(
+    (condition: V1NodeCondition) => condition.type === "Ready" && condition.status === "True",
+  );
+
+  const ip = getNodeInternalIP(node);
+  // in order to be added to the node set, the node must be ready. Additional
+  // status conditions will not cause the node to be removed however, e.g. disck
+  // pressure, memory pressure, etc. Nodes are only removed from the node set
+  // when they are deleted, and done so in the updateKubeNodesFromDelete function.
+  if (isReady) {
+    if (ip) nodeSet.add(ip);
+  }
+
+  await updateKubeNodesNetworkPolicies();
+}
+
+/**
+ * When a node is deleted, remove its IP from the set, rebuild the policies,
+ * and update the NetworkPolicies.
+ */
+export async function updateKubeNodesFromDelete(node: kind.Node) {
+  const ip = getNodeInternalIP(node);
+  if (ip) nodeSet.delete(ip);
+
+  await updateKubeNodesNetworkPolicies();
+}
+
+/**
+ * Fetch all Kubernetes nodes.
+ */
+async function fetchKubernetesNodes(): Promise<KubernetesListObject<kind.Node>> {
+  return K8s(kind.Node).Get();
+}
+
+/**
+ * Update all NetworkPolicies labeled with uds/generated=KubeNodes to
+ * reflect the given node CIDRs.
+ */
+export async function updateKubeNodesNetworkPolicies() {
+  const newNodes = buildNodePolicies([...nodeSet]);
+  const netPols = await K8s(kind.NetworkPolicy)
+    .WithLabel("uds/generated", RemoteGenerated.KubeNodes)
+    .Get();
+
+  for (const netPol of netPols.items) {
+    if (!netPol.spec) {
+      log.warn(
+        `KubeNodes NetworkPolicy ${netPol.metadata?.namespace}/${netPol.metadata?.name} is missing spec.`,
+      );
+      continue;
+    }
+
+    if (netPol.spec.egress) {
+      netPol.spec.egress[0] = netPol.spec.egress[0] || { to: [] };
+      netPol.spec.egress[0].to = newNodes;
+    } else if (netPol.spec.ingress) {
+      netPol.spec.ingress[0] = netPol.spec.ingress[0] || { from: [] };
+      netPol.spec.ingress[0].from = newNodes;
+    }
+
+    if (netPol.metadata) {
+      // Remove managed fields to prevent server-side apply errors
+      netPol.metadata.managedFields = undefined;
+    }
+
+    log.debug(
+      `Updating KubeNodes NetworkPolicy ${netPol.metadata?.namespace}/${netPol.metadata?.name} with new CIDRs.`,
+    );
+
+    try {
+      await K8s(kind.NetworkPolicy).Apply(netPol, { force: true });
```

Comment:
https://github.com/defenseunicorns/uds-core/blob/970-all-nodes/src/pepr/operator/controllers/network/generators/kubeAPI.ts#L173-L197

Would be good to compare with the existing node list before we make the apply to make sure we aren't unnecessarily calling k8s apply here.

Reply:
Changed in 9b7a61c3